### ISSUE TITLE
New lines have to be removed before entering SpaCy tokenizer

### ISSUE THREAD

**Issue by zzwx** [explosion/spaCy#7735]
## How to reproduce the behaviour

I'm bumping into an issue of tokenizing 2 similar sentences that only have new lines differences.
In a nutshell this is what I'm getting. As you can see the dep - head is completely screwed up. May be I'm not understanding it, but could that be a bug?

This one is using `spacy.load("en_core_web_trf")`: 

```
-----------------------------------------------------------------------------------------------------------------------
I am working third shift now.
-----------------------------------------------------------------------------------------------------------------------
 0: "I"                  "I"                  "i"                   PRON   PRP      "nsubj" [ 2]"working"              alpha:true ascii:true stop:true title:true
 1: "am"                 "be"                 "am"                   AUX   VBP        "aux" [ 2]"working"              alpha:true ascii:true stop:true
 2: "working"            "work"               "working"             VERB   VBG       "ROOT" [ 2]"working"              alpha:true ascii:true
 3: "third"              "third"              "third"                ADJ    JJ       "amod" [ 4]"shift"                alpha:true ascii:true stop:true like_num:true
 4: "shift"              "shift"              "shift"               NOUN    NN       "dobj" [ 2]"working"              alpha:true ascii:true
 5: "now"                "now"                "now"                  ADV    RB     "advmod" [ 2]"working"              alpha:true ascii:true stop:true
 6: "."                  "."                  "."                  PUNCT     .      "punct" [ 2]"working"              ascii:true punct:true
-----------------------------------------------------------------------------------------------------------------------
I am 
working third   

  shift now.
-----------------------------------------------------------------------------------------------------------------------
 0: "I"                  "I"                  "i"                   PRON   PRP      "nsubj" [ 2]"\n"                   alpha:true ascii:true stop:true title:true
 1: "am"                 "be"                 "am"                   AUX   VBP        "aux" [ 2]"\n"                   alpha:true ascii:true stop:true
 2: "\n"                 "\n"                 "\n"                 SPACE   _SP      "ccomp" [ 3]"working"              ascii:true space:true
 3: "working"            "work"               "working"             VERB   VBG       "ROOT" [ 3]"working"              alpha:true ascii:true
 4: "third"              "third"              "third"                ADJ    JJ       "amod" [ 5]"  \n\n  "             alpha:true ascii:true stop:true like_num:true
 5: "  \n\n  "           "  \n\n  "           "  \n\n  "           SPACE   _SP     "nummod" [ 6]"shift"                ascii:true space:true
 6: "shift"              "shift"              "shift"               NOUN    NN       "dobj" [ 3]"working"              alpha:true ascii:true
 7: "now"                "now"                "now"                  ADV    RB     "advmod" [ 3]"working"              alpha:true ascii:true stop:true
 8: "."                  "."                  "."                  PUNCT     .      "punct" [ 3]"working"              ascii:true punct:true
```

And now using `spacy.load("en_core_web_sm")`:

```
-----------------------------------------------------------------------------------------------------------------------
I am working third shift now.
-----------------------------------------------------------------------------------------------------------------------
 0: "I"                  "I"                  "i"                   PRON   PRP      "nsubj" [ 2]"working"              alpha:true ascii:true stop:true title:true
 1: "am"                 "be"                 "am"                   AUX   VBP        "aux" [ 2]"working"              alpha:true ascii:true stop:true
 2: "working"            "work"               "working"             VERB   VBG       "ROOT" [ 2]"working"              alpha:true ascii:true
 3: "third"              "third"              "third"                ADJ    JJ       "amod" [ 4]"shift"                alpha:true ascii:true stop:true like_num:true
 4: "shift"              "shift"              "shift"               NOUN    NN       "dobj" [ 2]"working"              alpha:true ascii:true
 5: "now"                "now"                "now"                  ADV    RB     "advmod" [ 2]"working"              alpha:true ascii:true stop:true
 6: "."                  "."                  "."                  PUNCT     .      "punct" [ 2]"working"              ascii:true punct:true
-----------------------------------------------------------------------------------------------------------------------
I am 
working third   

  shift now.
-----------------------------------------------------------------------------------------------------------------------
 0: "I"                  "I"                  "i"                   PRON   PRP      "nsubj" [ 3]"working"              alpha:true ascii:true stop:true title:true
 1: "am"                 "be"                 "am"                   AUX   VBP        "aux" [ 3]"working"              alpha:true ascii:true stop:true
 2: "\n"                 "\n"                 "\n"                 SPACE   _SP        "dep" [ 3]"working"              ascii:true space:true
 3: "working"            "work"               "working"             VERB   VBG       "ROOT" [ 3]"working"              alpha:true ascii:true
 4: "third"              "third"              "third"                ADV    RB     "advmod" [ 3]"working"              alpha:true ascii:true stop:true like_num:true
 5: "  \n\n  "           "  \n\n  "           "  \n\n  "           SPACE   _SP        "det" [ 6]"shift"                ascii:true space:true
 6: "shift"              "shift"              "shift"               NOUN    NN       "dobj" [ 3]"working"              alpha:true ascii:true
 7: "now"                "now"                "now"                  ADV    RB     "advmod" [ 3]"working"              alpha:true ascii:true stop:true
 8: "."                  "."                  "."                  PUNCT     .      "punct" [ 3]"working"              ascii:true punct:true
```

It seems that `SPACE` becomes a problem in tagging somehow.

## Your Environment

```
- **spaCy version:** 3.0.5
- **Platform:** Windows-10-10.0.19041-SP0
- **Python version:** 3.9.4
- **Pipelines:** en_core_web_sm (3.0.0), en_core_web_trf (3.0.0)
```

**Comment by polm** [comment]
That does look weird and may indicate a problem. The annotations look basically nonsensical. However, the main issue is probably that spaCy's training data doesn't have many (any?) examples of whitespace like this, so if you leave it in the results are basically undefined. It might be useful for textcat like application, but weird whitespace is one thing I would almost always recommend removing during preprocessing. 

If you're leaving it in for a specific reason could you tell us about your use case? It would give us a better idea of what the right behavior would be.

**Comment by zzwx** [comment]
Thank you for follow up.

My use case is to keep track of original text locations and the source is a markdown/commonmark consisting of text that has new lines to wrap long lines of text. So the example with `  \n\n  ` is most likely exaggerated. I'm expecting a single `\n` most of the time with a possibility of spaces, so the delimiter will consist of `<whitespace>+\n<whitespace>+` (if we assume whitespace has no `\n`).

I can definitely pre-sanitize, and upon receiving result back, retrace to original source using accumulated indices.

The problem is that even with single `\n` I get some strange results. For example for the `"I am\nworking third\nshift now."` input I get back **_two_** sentences, and this is using `spacy.load("en_core_web_trf")` model:

```
I am
working third
shift now.
-----------------------------------------------------------------------------------------------------------------------
 0: "I"                  "I"                  "i"                   PRON   PRP      "nsubj" [ 2]"\n"                   alpha:true ascii:true stop:true title:true
 1: "am"                 "be"                 "am"                   AUX   VBP        "aux" [ 2]"\n"                   alpha:true ascii:true stop:true
 2: "\n"                 "\n"                 "\n"                 SPACE   _SP      "ccomp" [ 3]"working"              ascii:true space:true
 3: "working"            "work"               "working"             VERB   VBG       "ROOT" [ 3]"working"              alpha:true ascii:true
 4: "third"              "third"              "third"                ADJ    JJ       "dobj" [ 3]"working"              alpha:true ascii:true stop:true like_num:true
-----------------------------------------------------------------------------------------------------------------------
 0: "\n"                 "\n"                 "\n"                 SPACE   _SP       "ROOT" [ 0]"\n"                   ascii:true space:true
 1: "shift"              "shift"              "shift"               NOUN    NN       "dobj" [ 0]"\n"                   alpha:true ascii:true
 2: "now"                "now"                "now"                  ADV    RB     "advmod" [ 0]"\n"                   alpha:true ascii:true stop:true
 3: "."                  "."                  "."                  PUNCT     .      "punct" [ 0]"\n"                   ascii:true punct:true
```

So the idea of pre-sanitizing is also lurking away

**Comment by polm** [comment]
Unfortunately single newlines are not better than double newlines here in the sense that I'm pretty sure they are also not present in the training data. 

Since in markdown single newlines are generally not different from spaces, I think it's easiest to just pre-sanitize your data, since you don't have much information to lose by doing so.

**Comment by zzwx** [comment]
I guess I was hoping that you'd say that new lines could be as well treated as space characters, and therefore whitespace recognition could be extended, but without knowing the internals and implications I understand this might be dangerous, is that right?

**Comment by polm** [comment]
I don't think it'd be dangerous, I think it just wouldn't work, or it would be equivalent to preprocessing your whitespace. 

spaCy's tokenization is designed to be non-destructive. One simplifying assumption it makes is that normal whitespace is just single ASCII spaces between tokens, the presence or absence of which is represented with a boolean token attribute. Anything not represented by that token attribute must be a token, so that using just the token surface (text) and boolean space attribute you can reconstruct the original string. There are other ways to represent tokenization, like indexes into the original string, but that's just not how spaCy works.

**Comment by zzwx** [comment]
Returning a token of whitespace that is not only consisting of single spaces but include \n or \r\n would greatly simplify the workflow honestly, otherwise the burden is on the caller to recognize and preprocess and then upon receiving the result to unwrap the accumulated offsets to match the original source character indices. 

If current recognition of the white space is a function of some kind, could it be exposed somehow as part of the API?

**Comment by polm** [comment]
I understand having more flexible whitespace handling would help you, but changing spaCy would create a binary compatibility issue with the Doc class, so it's not something we can do lightly, and most applications don't need this. 

If you want to see the code for whitespace handling it's [here](https://github.com/explosion/spaCy/blob/1d4b1dea25df6ab53521dc609d9b48a94385d1b0/spacy/tokenizer.pyx#L153). 

That said I think you may be overestimating the difficulty of creating an alignment mapping. You can take a look at [spacy-alignments](https://github.com/explosion/spacy-alignments), which handles much more tricky changes in sequence alignment. 

I think is is all you have to do:

```
import spacy

text = "I \t like\ncheese."

nlp = spacy.blank("en")

def make_aligned_doc(raw):
    #XXX replace all whitespace with spaces here
    doc = nlp(raw.replace("\n", " "))
    new2old = {}
    ii = 0 # index in new doc

    prev = None
    for tok in doc:
        if tok.is_space:
            continue
        
        new2old[ii] = tok.idx
        ii += len(tok.text_with_ws)

    text = [tok.text_with_ws for tok in doc if not tok.is_space]
    text = ''.join(text)

    return nlp(text), new2old

doc, alignment = make_aligned_doc(text)
print(doc)
print(alignment)
for (new, old), tok in zip(alignment.items(), doc):
    print(text[old:old+len(tok.text)])
```

**Comment by zzwx** [comment]
Thank you! I just renamed the issue in case anyone bumps into the same question. It can definitely be closed with "won't resolve" tag.

I ended up making a workaround in Go as this is where I'm calling SpaCy from. The workaround saves original locations before merging any spaces and after receiving tokens back, "unwraps" and adjusts the offsets. 

While this works OK, I'm still suggesting that SpaCy takes care of this step internally.

That could also include temporarily removing invalid unicode characters as they end up in an exception thrown. This can be argued depending on use cases, but I think this would end up more universal than unexpected sanitizing.
